subsys/net/conn_mgr/conn_mgr_monitor.c

```diff
@@ -91,6 +143,537 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
+					       strlen(tls_hostname));
+			if (ret < 0) {
+				LOG_ERR("setsockopt: %d", errno);
+				ret = -errno;
+				goto out;
+			}
+		}
+	}
+
+	addrlen = 0;
+
+	if (IS_ENABLED(CONFIG_NET_IPV4) && online_check->hostaddr.sa_family == AF_INET) {
+		addrlen = sizeof(struct sockaddr_in);
+	} else if (IS_ENABLED(CONFIG_NET_IPV6) && online_check->hostaddr.sa_family == AF_INET6) {
+		addrlen = sizeof(struct sockaddr_in6);
+	}
+
+	if (addrlen == 0) {
+		ret = -EINVAL;
+		NET_DBG("Invalid sockaddr len (%d)", ret);
+		goto out;
+	}
+
+	NET_DBG("Connecting to %s:%d",
+		online_check->hostaddr.sa_family == AF_INET ?
+		net_sprint_ipv4_addr(&net_sin(&online_check->hostaddr)->sin_addr) :
+		net_sprint_ipv6_addr(&net_sin6(&online_check->hostaddr)->sin6_addr),
+		online_check->hostaddr.sa_family == AF_INET ?
+		ntohs(net_sin(&online_check->hostaddr)->sin_port) :
+		ntohs(net_sin6(&online_check->hostaddr)->sin6_port));
+	ret = zsock_connect(sock, &online_check->hostaddr, addrlen);
+	if (ret < 0) {
+		ret = -errno;
+		NET_DBG("Connect failed (%d)", ret);
+		goto out;
+	}
+
+	NET_DBG("Sending HTTP%sGET request to %s:%d (ifindex %d)",
+		online_check->is_tls ? "S " : " ",
+		online_check->hostaddr.sa_family == AF_INET ?
+		net_sprint_ipv4_addr(&net_sin(&online_check->hostaddr)->sin_addr) :
+		net_sprint_ipv6_addr(&net_sin6(&online_check->hostaddr)->sin6_addr),
+		online_check->hostaddr.sa_family == AF_INET ?
+		ntohs(net_sin(&online_check->hostaddr)->sin_port) :
+		ntohs(net_sin6(&online_check->hostaddr)->sin6_port),
+		net_if_get_by_iface(iface));
+
+	exec_http_query(iface, sock);
+	ret = 0;
+
+out:
+	zsock_close(sock);
+	return ret;
+}
+
+static void do_online_check(struct net_if *iface)
+{
+	if (IS_ENABLED(CONFIG_NET_CONNECTION_MANAGER_ONLINE_CHECK_PING)) {
+		do_online_ping_check(iface, PING_HOST);
+	} else if (IS_ENABLED(CONFIG_NET_CONNECTION_MANAGER_ONLINE_CHECK_HTTP)) {
+		do_online_http_check(iface, URL);
+	}
+}
+
+static void online_connectivity_check(void)
+{
+	/* We create our own copy of the iface_states in order not to keep
+	 * the lock for long time. We also check from the master states
+	 * if the interface state has changed when launching the check for
+	 * that interface.
+	 */
+	static uint16_t states[CONN_MGR_IFACE_MAX];
+	int i;
+
+	if (!IS_ENABLED(CONFIG_NET_CONNECTION_MANAGER_ONLINE_CONNECTIVITY_CHECK)) {
+		return;
+	}
+
+	k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+	memcpy(states, iface_states, sizeof(states));
+	k_mutex_unlock(&conn_mgr_mon_lock);
+
+	for (i = 0; i < ARRAY_SIZE(states); i++) {
+		if (states[i] & CONN_MGR_IF_READY) {
+			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+			if (!(states[i] & CONN_MGR_IF_READY)) {
+				k_mutex_unlock(&conn_mgr_mon_lock);
+				continue;
+			}
+
+			k_mutex_unlock(&conn_mgr_mon_lock);
+
+			/* Do connectivity check and if it succeeds,
+			 * generate CONNECTIVITY_ONLINE event. The online
+			 * check might take some time.
+			 */
+			do_online_check(conn_mgr_mon_get_if_by_index(i));
+		}
+	}
+}
```

Comment:
> I don't think it is necessary to cache the states array, especially since you are locking the mutex briefly for ready ifaces anyways to check for readiness loss. At that point, I'd say just go for it and briefly check every iface, and save yourself the complexity.

I agree, it was a leftover when I had the mutex locked much longer, I will remove the cache.